src/gvars.c

```diff
@@ -226,6 +226,67 @@ static Obj             FopiesGVars;
 
 #endif
 
+// FlagsGVars contains information about global variables as a single INTOBJ
+// Once turned into a C int, this information is:
+//
+// Bits 0 and 1 - A value of type GVarWriteFlag which denotes if the variable
+// is ReadOnly, Mutable or Constant
+// Bit 2 - If the variable has ever had a non-default value assigned to
+// ExprGVars,
+//         CopiesGVars or FopiesGVars. Note that this bit is never cleared at
+//         present, so it can be set to 1 for a gvar, while these three arrays
+//         all have their default value.
+
+// Below are some helper methods for reading and writing FlagsGVars. They
+// should always be used to simplifying adding more flags in future.
+
+typedef enum {
+    GVarReadOnly = 0,
+    GVarMutable = 1,
+    GVarConstant = 2
+} GVarWriteFlag;
+
+static GVarWriteFlag GetGVarWriteState(Int gvar)
+{
+    UInt val = INT_INTOBJ(ELM_GVAR_LIST(FlagsGVars, gvar));
+    return val & 3;
+}
+
+typedef struct {
+    GVarWriteFlag gvarWriteFlag;
+    Int           hasExprCopiesFopies;
+} GVarFlagInfo;
+
+static GVarFlagInfo GetGVarFlagInfo(Int gvar)
+{
+    UInt         val = INT_INTOBJ(ELM_GVAR_LIST(FlagsGVars, gvar));
+    GVarFlagInfo info;
+    info.gvarWriteFlag = val & 3;
+    info.hasExprCopiesFopies = val & 4;
+    return info;
+}
+
+static void SetGVarWriteState(Int gvar, GVarWriteFlag w)
+{
+    UInt val = INT_INTOBJ(ELM_GVAR_LIST(FlagsGVars, gvar));
```

Comment:
When this is called the very first time, `ELM_GVAR_LIST(FlagsGVars, gvar)` returns `0`, and thus an assertion in `INT_INTOBJ` is triggered. So we need to deal with that case explicitly. Also in the other places accessing this.
Alternatively, we could just change the initialisation of `FlagsGVars` to set it to `INTOBJ_INT(0)` at first.